state-chain/generic-typeinfo-derive/src/lib.rs

```diff
@@ -0,0 +1,110 @@
+extern crate proc_macro;
+use proc_macro::TokenStream;
+use proc_macro2::TokenStream as TokenStream2;
+use quote::quote;
+use syn::{Data, DataEnum, DataStruct, DeriveInput, Fields, Ident, ImplGenerics, TypeGenerics};
+
+#[proc_macro_derive(GenericTypeInfo)]
+pub fn derive(item: TokenStream) -> TokenStream {
+	let item2: TokenStream2 = item.into();
+	let ast: DeriveInput = syn::parse2(item2).unwrap();
+	let ident = ast.ident;
+	let (impl_generics, ty_generics, _) = ast.generics.split_for_impl();
+	match ast.data {
+		Data::Struct(d) => derive_struct(d, ident, impl_generics, ty_generics),
+		Data::Enum(d) => derive_enum(d, ident, impl_generics, ty_generics),
+		Data::Union(_) => panic!("#[derive(GenericTypeInfo)] not yet implemented for Unions"),
+	}
+	.into()
+}
+
+fn derive_fields(d: Fields) -> TokenStream2 {
+	match d {
+		syn::Fields::Named(fields_named) => {
+			let fields: Vec<TokenStream2> = fields_named.named.iter().map(|field| {
+                let ty = field.clone().ty;
+                let typename = clean_type_string(&quote!(#ty).to_string());
+                let name = field.clone().ident.unwrap();
+                quote!(.field(|f| {f.ty::<#ty>().type_name(scale_info::prelude::format!("{}For{}", #typename, T::NAME).leak()).name(::core::stringify!(#name))}))
```

Comment:
Am I right that this assumes that the type of the field *also* has `#[derive(GenericTypeInfo)]`? 

For example this code:

```
#[derive(GenericTypeInfo)]
struct A<T, I> {
	inner: B<T, I>,
}
```

The inner type will be set to `BForT` - but what if we don't derive GenericTypeInfo for B? Would this be caught at compile time? Or does it somehow propagate magically through the typeinfo registry?


(Side not - tbh. I would also prefer some more explicit distinguishing constant than just `NAME`. Like `TYPE_INFO_SUFFIX` or something.)

Reply:
Also btw. we're assuming that the type param will always be `T` - probably ok for now but we can probably also pull the actual type param from the parsed input.